src/common/CryptoUtils.ts

```diff
@@ -0,0 +1,150 @@
+            if (diff < millisecondsInOneDay || diff > 825 * millisecondsInOneDay) {
+                throw new Error(messages.getMessage('error:invalidValidityDate'));
+            }
+        } else if (validity < 1 || validity > 825) {
+            throw new Error(messages.getMessage('error:invalidValidityNumber'));
+        }
+
+        const keys = forge.pki.rsa.generateKeyPair(keySize);
+        const cert = forge.pki.createCertificate();
+
+        const startDate = new Date();
+        let endDate = new Date();
+        if (validity instanceof Date) {
+            endDate = validity;
+        } else {
+            endDate.setDate(startDate.getDate() + validity);
+        }
+
+        cert.publicKey = keys.publicKey;
+        cert.serialNumber = '01';
+        cert.validity.notBefore = startDate;
+        cert.validity.notAfter = endDate;
+
+        const attrs = [
+            {
+                name: 'commonName',
+                value: hostname
+            },
+            {
+                name: 'countryName',
+                value: 'US'
+            },
+            {
+                shortName: 'ST',
+                value: 'California'
+            },
+            {
+                name: 'localityName',
+                value: 'San Francisco'
+            },
+            {
+                name: 'organizationName',
+                value: 'Example Inc.'
+            },
+            {
+                shortName: 'OU',
+                value: 'Test'
+            }
+        ];
+
+        cert.setSubject(attrs);
+        cert.setIssuer(attrs);
+
+        // Add the subjectAltName and ExtendedKeyUsage extensions, which are required by Apple
+        cert.setExtensions([
+            {
+                name: 'basicConstraints',
+                cA: true
+            },
+            {
+                name: 'keyUsage',
+                keyCertSign: true,
+                digitalSignature: true,
+                nonRepudiation: true,
+                keyEncipherment: true,
+                dataEncipherment: true
+            },
+            {
+                name: 'extKeyUsage',
+                serverAuth: true,
+                clientAuth: true,
+                codeSigning: true,
+                emailProtection: true,
+                timeStamping: true
+            },
+            {
+                name: 'subjectAltName',
+                altNames: [
+                    {
+                        type: 2, // DNS
+                        value: hostname
+                    },
+                    {
+                        type: 7, // IP
+                        value: '127.0.0.1'
+                    },
+                    {
+                        type: 7, // IP
+                        value: '10.0.2.2'
+                    },
+                    {
+                        type: 7, // IP (IPv6 loopback)
+                        value: '::1'
+                    }
```

Comment:
regardless of whether `localhost` is being used for `hostname` or not, I think we should always white-list these special IPs..... any objections?

Reply:
I think this is a good idea, and basically kills two birds with one stone with respect to Android emulators. Works for me.